integration/e2e_test.go

```diff
@@ -0,0 +1,230 @@
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	gcpcachev2 "github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	gcpserverv2 "github.com/envoyproxy/go-control-plane/pkg/server/v2"
+	gcpserverv3 "github.com/envoyproxy/go-control-plane/pkg/server/v3"
+	gcptest "github.com/envoyproxy/go-control-plane/pkg/test"
+	gcpresourcev2 "github.com/envoyproxy/go-control-plane/pkg/test/resource/v2"
+	gcptestv2 "github.com/envoyproxy/go-control-plane/pkg/test/v2"
+	"github.com/envoyproxy/xds-relay/internal/app/server"
+	yamlproto "github.com/envoyproxy/xds-relay/internal/pkg/util"
+	aggregationv1 "github.com/envoyproxy/xds-relay/pkg/api/aggregation/v1"
+	bootstrapv1 "github.com/envoyproxy/xds-relay/pkg/api/bootstrap/v1"
+	"github.com/onsi/gomega"
+)
+
+func TestMain(m *testing.M) {
+	// We force a 1 second sleep before running a test to let the OS close any lingering socket from previous
+	// tests.
+	time.Sleep(1 * time.Second)
+	code := m.Run()
+	os.Exit(code)
+}
+
+func TestSnapshotCacheSingleEnvoyAndXdsRelayServer(t *testing.T) {
+	if runtime.GOOS != "linux" {
+		t.Skip("Golang does not offer a cross-platform safe way of killing child processes, so we skip these tests if not on linux.")
+	}
+
+	g := gomega.NewWithT(t)
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	// Test parameters
+	var port uint = 18000
+	var upstreamPort uint = 18080
+	var basePort uint = 9000
+	var nClusters = 7
+	var nListeners = 9
+	var nUpdates = 4
+
+	// We run a service that returns the string "Hi, there!" locally and expose it through envoy.
+	go gcptest.RunHTTP(ctx, upstreamPort)
+
+	configv2, snapshotv2, signal := startSnapshotCache(ctx, upstreamPort, basePort, nClusters, nListeners, port)
+
+	// Start xds-relay server. Note that we are starting the server now but the envoy instance is not yet
+	// connecting to it since the orchestrator implementation is still a work in progress.
+	// TODO: parametrize configuration files
+	startXdsRelayServer(ctx, "./testdata/bootstrap_configuration_e2e.yaml", "./testdata/keyer_configuration_complete_tech_spec.yaml")
+
+	// Start envoy and return a bytes buffer containing the envoy logs
+	// TODO: parametrize bootstrap file
+	// TODO: hook up envoy to the xds-relay server
+	envoyLogsBuffer := startEnvoy(ctx, "./testdata/envoy_bootstrap.yaml", signal)
+
+	for i := 0; i < nUpdates; i++ {
+		snapshotv2.Version = fmt.Sprintf("v%d", i)
+		log.Printf("Update snapshot %v\n", snapshotv2.Version)
+
+		snapshotv2 := snapshotv2.Generate()
+		if err := snapshotv2.Consistent(); err != nil {
+			log.Printf("Snapshot inconsistency: %+v\n", snapshotv2)
+		}
+
+		// TODO: parametrize node-id in bootstrap files.
+		err := configv2.SetSnapshot("test-id", snapshotv2)
+		if err != nil {
+			t.Fatalf("Snapshot error %q for %+v\n", err, snapshotv2)
+		}
+
+		g.Eventually(func() (int, int) {
+			ok, failed := callLocalService(basePort, nListeners)
```

Comment:
we can probably avoid setting up and calling a local service and instead use envoy stats to check config was applied. That way we can probably remove dependency on `Eventually`. wdyt?

Reply:
we'd still have a similar problem, right? As in, at which point can we check the `/stats` enpdoint?

Reply:
yeah i see..`Eventually` still has to be used. can we still use envoy stats to verify instead of a local service.

Reply:
as i think more, i think what we have is better since it validates lds/cds/rds/eds were configured right.